### ISSUE TITLE
Third-party API returns malformed A2S results

### ISSUE THREAD

**Issue by aclist** [aclist/dztui#13]
API may rarely return improperly byte-decoded modlist and mangle the output, concatenating multiple mods into a non-existent mod and assigning the wrong publishedfileid of some other content.

**Comment by aclist** [comment]
Cf. https://www.battlemetrics.com/servers/dayz/13328284

**Comment by SteveIDusa** [comment]
### @aclist 

As directed, I will now add a 3rd Server that no longer works: 13193240.  This is the one that I have been playing with my Wife for last few weeks with out any issue.  Logged several hundred hours with here in fact since finding DZGUI.  But all of a sudden it too has the malformed character and links.

I am beginning to wonder if some one is being malicious....this was working perfectly!

**Comment by aclist** [comment]
Preparing a hotfix...they are forcing my hand. It is working, but need to test that the game launches correctly.

Yeah, the modlist for your server is a complete mess now.

**Comment by aclist** [comment]
@SteveIDusa 

Here is a hotfix, please test: https://raw.githubusercontent.com/aclist/dztui/testing/dzgui.sh

This is outside of the auto-download functionality, so you'll have to save the contents into a text file, make it executable, then run it. I disabled the version check, so it won't prompt you to fetch a different version.

Once this is working, I'll push it to the main branch. (Maybe there should be a mechanism for downloading different versions from inside the app, (e.g. stable/testing branch) but seems kind of overengineered.)

Thanks.

**Comment by SteveIDusa** [comment]
### @aclist 
I had noticed the name for 15082587, I only included it as it was in the top 10 and I was just really taking a look at if the problem was prolific or not.  It seems it is prolific or is either becoming prolific, so nice boon that it made you consider something perhaps not really considered previously since it was not part of the problem that you are addressing.

Regarding the Hotfix, do I need to place this inside the folder with my other dzgui downloads or place it somewhere else?

**Comment by SteveIDusa** [comment]
### @aclist 
I dropped it into a folder adjacent to my original downloads, saved it as a .sh as the others and set to executable.  When run the Newer Version available dialog does pop up which shows that my version is 2.2.1-rc.1 and wants to point me to 2.2.0.  I have told it No.

The last thing that I was doing last night was removing mods to try again when I realized I was to tired to continue, so this did cause DZGUI to prompt to download the mods.  Which was likely a good test...

Oh.....trying to download via Steam Client rather than the website?......Nice!  Slower, but nice! No need to sign into Steam on their website. 

I really wish Steam would update the Steam Client to show you what is downloading from the Workshop for DayZ.  It just says DayZ and stacks them all up.  Is it like this for all games and the Workshop?  That is irrelevant and you do not need to address that.

All of the mods downloaded as expected after clicking the subscribe button.

Loaded into the game without an issue!

Thank you!  What is different?  What crossed your mind to see a way around BattleMetrics?

**Comment by aclist** [comment]
Sorry, I think I intended to comment out the version check, but then I made some changes, and I ended up overwriting everything again. Anyway, I pushed this version to the mainline branch, so you shouldn't get a version conflict popup anymore, unless you use the old file. Happy to hear it is working.

Yeah, I noticed we were listing the web site URLs and just figured might as well use the internal Steam URLs instead. I think they just queue up the mods under the games in the downloads page. I don't know if you can expand the info on that page, never tried.

I'll outline what was changed tomorrow, kind of tired right now. I also started experimenting a bit with a method that would use Steam native APIs to get server status, fetch mod list, and find IDs of mods, but it's a bit flaky at the moment and I need to ensure it's reliable. If we go that route, it would mean the user has to generate a Steam API key instead of a BM one, but it should be much more reliable than a third party service.

**Comment by SteveIDusa** [comment]
Sounds good, looking forward to it, but no hurry. Sounds like you could use some rest!  Thank you again!

**Comment by aclist** [comment]
It looks like I introduced a regression having to do with the fav server select during the hasty hotfix (https://github.com/aclist/dztui/issues/17). I just pushed a new fix for that. I believe that probably relates to the issue you described, but I was a bit unclear from your description what exact steps led to it occurring. It sounds like a variation of an old bug where when you launched a server from the main menu, then exited back, then selected favorite, it was sourcing the parameters from the first server instead of updating.

Try the new version and see if it goes away. There shouldn't be any situations where the main server select leaks into the fav server at this point, so far as I can see.

As for the fix, it's definitely a workaround and not a beautiful solution. Now we're using Yet Another Third-Party API to fetch the correct mod IDs, and only using BM to get the basic server info (since their mod info can no longer be relied upon--"fool me once"). This is fine, but the weakest link is still these third party APIs. It looks like Valve's first-party API lets you get all of the server information we are currently getting, but when it comes to the mods, only returns the literal names reported by the server, so would have to do a reverse lookup on the names using a different Valve interface to find their actual IDs (for downloading). 

This all seems to work OK from testing, but getting the proper names back from the server encoded without any garbage is a bit problematic still. I did also find a method of getting the list of every DayZ server reporting as public (10k+ in one shot) without a third party API, so we could theoretically even make a full-scale server browser, but that seems totally unnecessary for now. I guess the basic workflow would be that you'd used the BM web site to browse for servers you like (lots of info about uptime, proximity, etc. that is probably more useful than a generic server browser I could make), then use the IP/port as an identifier instead of the server ID, and process the rest directly through the Steam APIs. I need to make a proof of concept and see how it all works end to end.

Numbered mod list: I added it.

I have introduced a method of toggling between stable/testing branches of DZGUI from within the app. I want to use this to deploy new features "over the air" in testing without affecting other users. 

If you toggle debug mode from the main menu, then enter the debug options mode, this exposes a "toggle branch" feature. This lets you switch to the testing branch, and will use the existing download logic to pull down the latest version if there is a new one. You can switch between branches on the fly and it'll always check whether you are in synch.

I'd like to start using that to trial fully automatic downloading of mods, since I think I finally conceptualized a way of using steamcmd that is going to be low-impact for the user. Downloading 20 mods is extremely annoying, particularly when troubleshooting, so I'd like to see if this can be automated while still being user-friendly. Fully automated downloading is already working on DZTUI, so it's just bringing that functionality over while making it a bit more intuitive. If you want to be a guinea pig for that, I'll start working on that and release it to the testing branch when ready.

Currently, both stable and testing are running the same version.

Priority for now is:
1. Fix some more stuff related to mod symlinks
2. Auto mod downloads
3. Prototype first-party API pipeline without BM

**Comment by SteveIDusa** [comment]
### @aclist 
Got home this evening and deleted all of my mods, unsubscribed and deleted my config.  Launched DZGUI and everything seems to work as expected.  Numbered list is a nice touch, thank you!  Did not remember however to us the "Toggle Branch" Feature, will take a look at it tomorrow evening.  Looking great so far!

**Comment by SteveIDusa** [comment]
Toggled Debug and found the Change Branch, though I did not see any changes take place, even restarting DZGUI did not prompt to download anything.  But was able again to login to server and play for a few hours with the wife.  Let me know if I missed anything or if you want anything tested specifically.  I have read the above post a couple of time but it always after I am tired and ready to unwind...lol.  I may be missing things.

**Comment by aclist** [comment]
It's nice of you to always go the extra mile and clear out your mods and reinstall them when testing.

The stable and testing branch are currently on the same version, so even if you toggle to testing, there will be nothing new to download--yet. When new features are elaborated on testing, they can be grabbed by switching to that branch. 

This is just a backend change that will facilitate introducing unstable features without affecting the rest of the population.

**Comment by aclist** [comment]
@SteveIDusa 

Hi, hope you are having a good weekend.

I pushed a release candidate to the testing branch that contains a large number of small features. If you switch to testing and update versions to `2.4.0-rc.1`, you should see numerous quality of life changes listed in the changelog. 

Ones of note are:

- Over-the-air news updates: used to push critical news outside of the regular release cadence. Once a message is seen by the client, it will be suppressed on next restart and the news header hidden.
- Prompt user to permanently update map count value
- Ability to delete servers from the list
- Better handling of servers with enormous numbers of mods

The last one is important. It changes the way symlinks are created by encoding them into shorter IDs, instead of verbatim mod names. There were some servers with so many mods, that the launch options are too long for Steam to handle, and the mod names have to be minified. 

To counteract the names becoming more obscure, the List Installed Mods option now also displays the corresponding symlink ID next to the real mod name. The user shouldn't need to touch the symlinks under normal use, but it's there just in case.

This last part should be working, but it does affect the way we prepare the launch options for Steam, so please test it. You should also see that in your DayZ path, the symlinks now have short alphanumeric names. It should also be smart enough to update your existing mod links to the new format, so I would recommend testing in two ways:

- Test that you can continue to connect to old servers out of the box
- Test that you can connect to servers when newly fetching mods

Thanks in advance.

**Comment by SteveIDusa** [comment]
### @aclist 
Got a chance to try the new rc today, sorry about the delay, family came down with covid.  Was able to connect to old server with out issue today.  Will follow up regarding testing fetch of mods in the next couple of days.

**Comment by aclist** [comment]
Sorry to hear that. Hope you will feel better soon.

**Comment by SteveIDusa** [comment]
### @aclist 
Had a chance to test further today, added a third server that needed mods downloaded.  Seemed to work just fine and only added missing mods.  One thing that I did notice is that every time I click on the DZGUI generated html page with the links and click on the first link for the first time, it always asks me if I want to have my web browser to always remember my choice, but it only does it for that session.  Next time I have to download mods it will do that again.  Is this an issue with my web browser?

**Comment by aclist** [comment]
1. Could you list the server this was on? If you still have the mods installed, could you check the output of the "List installed mods" command and look for the mods from that server to see if anything looks different?
2. About the browser prompt, you're referring to when the browser asks you if you want to associate `steam://` links as always openable in Steam, correct? That's a browser-side issue. I'm not sure why it's not persisting, but I can test it. What browser are you using?
3. You downgraded to Stable using the internal toggle in the app, right?

**Comment by SteveIDusa** [comment]
### @aclist 

1. Server ID# is 12528079,  Yes Mods are still installed, name of Mod that was displayed in the error is ZDS_Signs.  I even thought to check battlemetrics for that server after getting the error and the mod list was very clean looking to me.  However another thing that I have noticed since is that for this server and only this server of my choosen 3 the Ping is saying Timeout rather than giving a number....So perhaps it was a false positive???  But I did try twice using the RC version before trying the stable version, and this morning the mod list provided by DZGUI is looking just fine.....
2. Yes, the browser prompt is exactly what I was referring to.  No worries if it is a browser side thing,  Browser being used is Firefox, version 103.0, but it is customized by the Mint Distribution team, so I doubt this is something that you really need to worry about.
3. No, I have been getting a prompt on launch that a new version is available (though I think I would call it a regression considering it is going backwards..).  I have been telling it to not update, and in my current state my brain said, "Well, just restart it and tell it to grab the other version.", which then it indicated it downloaded and made a new config file.  Then it worked, without even prompting for a new download of the mod.   Sorry, the in app toggle did not even cross my mind....

**Comment by aclist** [comment]
1. OK, I'll test with that server. I don't think timeout would affect it. We're symlinking the mods in a different format in the release candidate (RC) version, which is why it's important to see that nothing is breaking on that front.
2. Hopefully this becomes a non-issue once I elaborate a method of auto downloading mods.
3. OK, switching versions is fully available in the Stable version as well, so you shouldn't have to manually fetch versions anymore. It works like this:

- Choose "Toggle debug mode"
- Opens a "Debug options" on the main menu
- Enter "Debug options"
- Choose "Toggle branch" (toggles between Stable/Testing)
- This updates the "branch" key on the config file
- Remember that you are in Debug mode still, so to actually connect to servers, choose "Toggle debug mode" again to revert to Normal mode (note the header)

This is done deliberately to prevent the casual user from switching to the Testing (unstable) branch unless they explicitly open the debug options.

As for versioning, it works like this:

If you are on Stable => checks upstream Stable version for changes
If you are on Testing => checks upstream Testing version for changes

When you switch between versions, you will always receive a warning to update, because the Testing version by definition won't be the same as the Stable version. Each time you switch branches, it just fast updates the local file with one or the other version. There is no management of versions locally, so you will always have one or the other version fetched directly from upstream.

As for the message itself, it's a bit of a misnomer to say "newer version" here, since the version check is merely checking whether local version ==  upstream version, in order to ensure total parity. It's never going to be the case that the upstream version is a lower version number (patches increment the version), so except in the case you mentioned, "newer version" should essentially be a true statement.

If your config is currently on Stable and you manually download a different branch from GitHub without updating your config file, such as (e.g.):

- Your config file points to Stable, Stable is on version 39
- You download version 49 from Testing branch without updating to Testing branch
- Your version must be 39, but it is 49
- Message tells you "new version" is out

It would then appear that the "newer version" message is incorrect, since the app will be checking the Stable branch, seeing a version mismatch, then prompting you to "update." But what it is in fact doing is prompting you to synchronize your stated branch in your config file with the master version on that branch. 

```
Newer version available.

Your branch:			stable
Your version:			2.4.0-rc.1
Upstream version:		2.3.0
```

Note the `Your branch` message when receiving an update warning. Perhaps the key hint you missed here is that an `-rc` version can never appear in `stable`, since those are testing versions. The message looks like a downgrade, but it's in fact a version conflict.

To alleviate these kinds of version conflicts, the in-app branch toggle always updates both the config file and the version, so you can switch between either branch and never fall out of synch. 

I suggest you use the internal toggle and always accept the download when switching branches.

I'm elaborating on this because I think I will reformulate the warning to be clearer:

```
Version conflict:

Your branch:			stable
Your version:			2.4.0-rc.1
Upstream version:		2.3.0
```

But in practice, this should never happen if using the in-app toggle. I also asked if you had used the in-app toggle because I wanted to know whether I made it clear enough to use, but it sounds like I hid it a bit too much (for the aforementioned reasons of not letting users hop into the testing branch unless they really want to.)

The one flaw here I can see is that if someone toggles to Testing, then toggles back to Stable, but decides to refuse the download (as you did), they are going to run into possible problems, since they won't actually be on the Stable version. I think what needs to happen here is that the download prompt when switching versions is enforced. That's an oversight on my part, will fix it.

Once you get up to speed on the in-app toggle and the versioning, you could try checking the server in question again, making note of which branch you are on.

**Comment by aclist** [comment]
OK, I tested the server in question and wasn't able to reproduce your issue. Sounds like a false positive. Was able to get into the server. 

The testing branch uses a new method of creating symlinks, while the stable version is on the old format. When using the new version, it checks for the "legacy" version of symlinks and updates them all to the new format.

Since you mentioned only downloading some missing mods, I downloaded some of the mods for that server on the stable version, then the remainder on the new version, to see if I could trigger a conflict in the way the symlinks are being prepared, but the additional methods I added for handling legacy symlinks in the testing branch ensured that everything was created correctly. For that matter, even if you had downloaded mods with the new version first, the old version also takes care to prepare all of the mods in its old format, so I don't think there is a scenario where they could get misaligned. I added a lot of precautions to ensure that whatever version you are using, the mods don't break.

I'm working up a series of changes over on testing:

- Enforce downloads when switching branches, don't prompt for it. (Ensures a branch toggle triggers fetching the correct version)
- Update config values better when toggling
- Continue prompting for normal downloads when booting up (new version available on your current branch)

This won't be totally seamless until these changes are propagated to Stable with version `2.4.0`, which I think is basically ready to go, but I'll await your confirmation. Thanks.

**Comment by SteveIDusa** [comment]
### @aclist 
You are doing a great deal of work!  Thank you!

Had an opportunity today to switch to debug mode, then changed branch to testing.  Attempted to connect to the offending server again while in debug.  Received the dry run message with launch options and clicked on the write to file button.

Then went back into debug to turn off debug mode and launched the offending server again.  This time it launched DayZ but again gave the Warning (0x00040073) message and stating that client was kicked from game due to data verification error, Client missing a Mod which is on the server.  Again indicating that the (ZDS_Signs) mod is missing with the Missing PBO. message.

I then closed DayZ and relaunched DZGUI, on this launch I did not get a message about downloading a different version it just automatically loaded the testing branch.  I then again toggled debug mode, choose the debug options and toggled branch back out of testing.  This then caused DZGUI to display the "DZGUI 2.3.0 successfully downloaded message.

I then relaunched DZGUI again, which indicated it wrote a new config file.  I then attempted to join the offending server again, this time out of the testing branch and on the stable branch.  With this launch of DayZ however I now received a Connecting Failed Error (0x00020010) error "This server is currently locked" and not the missing .pbo error as with the testing branch.  I then gave it a few moments and tried again, just to ensure that perhaps I had logged on just as the server was doing a restart or something.  But received another connecting failed error (slightly different number) with could not connect as server is unreachable error.  Though my wifes windows machine could see it at the same time.  

So I don't know......

**Comment by aclist** [comment]
Thank you for the testing!

Writing the dry run launch options to a file is intended to let you keep a log of it in case you want to share the exact launch parameters with the developer. Is this clear? Maybe the dialog needs a reword.

It honestly sounds like you got unlucky with the timing and hit this server just as they were locking/unlocking it and switching some stuff around (that's fairly common). A server might show a current player count but be locked to new joiners.

I'm a bit concerned about that particular mod causing an error for you, but it is true that some servers enforce a specific load order, so if it was in the midst of some reset, perhaps something funny happened.

I'm not discounting the possibility that this is a bug, but I haven't been able to reproduce any connection errors on any servers yet. 

I would greatly appreciate it if you could, when you have time, attempt to connect to this same server while on Testing and repeat the test 3-5 times. If it always happens, I will consider it demonstrable proof, and we can focus on tracking down what is going on with your setup.

**Comment by aclist** [comment]
It looks like Bohemia finally provided some additional info on how to decode the additional mod info they are sending over the A2S protocol, so Battlemetrics is not showing malformed IDs now. Looks like we can revert to the old method.

**Comment by SteveIDusa** [comment]
### @aclist 

Emerged from a fog, the light is bright!  

Regarding the launch options, it was clear to me that it was for providing to you.  However it did not include a location as to where it was written, so I had no idea where to find it.

That was likely it for the server being locked to new joiners, my wife recently had the experience of playing while others could not join until they updated a particular mod.  She came to the conclusion after the fact that had she disconnected during that time frame she would not have been able to rejoin until she updated the mod.  She came to this conclusion after finding that the mod had updated during her play.

I also have not had the mod issue after that last one, so I think you were right, likely not a bug but rather timing.  Unfortunately, It has been long enough now that I can not remember which server was having the issue.

I did note that you have done a couple of updates while I have been "mindless".  The last one and the QOL changes are very nice!  I saw the changes to the size and shape of the dialog boxes and I for one really appreciated these changes!  I am also now fully aware that dialog boxes that display a sentence as a sentence rather than a stack of words is something that can be changed.  I will never look at another Linux program the same way again....you have raised the bar!

I also noted that during my mindless phase, that at some point I stopped opening a terminal and setting the map size....and I have never made the change permanent, so am I right to think that it crossed your mind to add this to the launch options so that it just happens for any user?  Either way, it is nice to be mindless when launching a game!  So if you found a way to do it or if Bohemia made a change I like it!

So I am way out of the loop now, and behind in your progress.  So, is there anything you would like me to test or stress?

**Comment by aclist** [comment]
Hey, glad you are feeling better. 

Just the man I wanted to see! People keep submitting reports and then leaving after the bug is fixed without confirming whether everything worked as expected. (Which I guess means it is?)

> Regarding the launch options, it was clear to me that it was for providing to you. However it did not include a location as to where it was written, so I had no idea where to find it.

OK, right, I can add that.

> I am also now fully aware that dialog boxes that display a sentence as a sentence rather than a stack of words is something that can be changed. I will never look at another Linux program the same way again....you have raised the bar!

Not sure if you are making fun of me or not, lol. I noticed when testing on Steam Deck that some dialog boxes were getting super squashed, so it needed to be fixed. There may be some other dialogs that don't look right, so if you see any, let me know.

> I also noted that during my mindless phase, that at some point I stopped opening a terminal and setting the map size....and I have never made the change permanent, so am I right to think that it crossed your mind to add this to the launch options so that it just happens for any user? Either way, it is nice to be mindless when launching a game! So if you found a way to do it or if Bohemia made a change I like it!

The app includes a check now at start and will prompt the user to set the map count if it's too low. It might have changed on the Proton end, I don't know. You can check your current count by invoking `sysctl -q vm.max_map_count | awk -F"= " '{print $2}'`

> So I am way out of the loop now, and behind in your progress. So, is there anything you would like me to test or stress? I am currently on 2.4.1 Stable Branch.

There's a new pending release candidate on Testing that includes fixes for two bugs:
1. Bug that was causing having 10+ servers in the list to draw the table incorrectly => can now add arbitrary number of servers
2. Issue with mod collisions due to symlinks having similar names => changed symlink logic to prevent this

> Followed the instructions for the artwork and everything except the icon was working perfectly within steam. Took a couple of minutes to fumble my way to it, but I finally realized I could change the file type from .png to all files in the dialog box and get the icon image to populate in Steam. Looks great!

Right, I should add a note to select All Files.

> After playing for a bit last night, got up this morning, responded here and attempted to launch the game. Was immediately hit with missing PBO error for X2Taser (1593012445), which was not happening last night on server ID# 13193240. Attempted to unsubscribe from this mod and then re-subscribe to same result. Then unsubscribed from all mods, deleted all references on system both in .local and .steam locations. After this did a local file verification and fixed missing files, did not realize that I should not delete the "bliss" mod. Verification got it back and it showed as only mod installed. (This must be a bohemia provided mod?)

Not sure about the bliss mod or what it is. I suggest trying the Testing branch because of bugfix 2 listed above. There might be a collision in the symlinks. 

By the way, when troubleshooting mods, you can turn on Debug mode, go into Debug options, and select "Generate debug log," and it'll write a log file containing all of the mod details and other diagnostic data.

> I grabbed a screen shot but do not know how to attach it here.

I believe you can just drag/drop images directly into the text pane and it'll upload them.

**Comment by SteveIDusa** [comment]
### @aclist 

Not making fun of at all!  You literally have raised the bar by which I will judge other software dialog boxes.  I thought it was a linux thing for years, that it was something that the programmers had no control over.  I think this no more, and as I become more proficient with programming, it will effect how I approach interfaces.

Speaking of interfaces,  I got so fed up with my Linux Mint installation today I decided to backup what was important, nuke the drive and do a fresh install of Linux Mint 21 Mate Desktop Edition.  This gave me a chance to download the DZGU .zip to a new installation.  It has taken me all day to get DayZ up and running again.  For some reason, the installation paths changed and are no longer the same as the last installation.

Previously, I had the .local/Steam/  File path, now I have /.steam/debian-installation/steamapps/common/DayZ

The DZGUI install.sh script, I set to executable, and ran in a terminal, it allows me to populate a name, the API Key,  a server ID# and then immediately jumps to a Searching for alternate DayZ path progress bar.  (by the way, I see a typo....it says "Searching for alternate DayZ path (make take some time)....should that be "may take some time"?

Unfortunately, it has been here for quite a while and has not found the path above.....activity light seems to indicate it is doing something , but I can not believe it would take this much time.....perhaps I am not being patient enough....would it be possible to have it ask if the user would like to specify the patch or do an automatic search?

The only thing that I can think of that I did different, was to download steam via the built in Mint Software Store, rather than getting it from steam directly on their website.....should I consider nuking the installation and grabbing steam official?

**Comment by aclist** [comment]
Are you talking about this one?

`Launch conditions satisfied. DayZ will now launch after clicking [OK].`

There's a lot of variability in how different distributions and desktop environments render these things, so the best we can do is err on the side of generous sizing and padding and hope it looks decent in most cases. I was surprised to see that a lot of dialogs on Steam Deck were crammed into thin strips in the center of the screen (taking up 5% of the screen) until I changed the dimensions.

The path search takes quite a long while on my machine as well, since it scans literally every directory. Try waiting and seeing if it takes longer than 5 minutes? We have no guarantee of where someone could have installed the game, so I thought this was safer. But maybe certain paths it absolutely can't be should be excluded altogether. But maybe this is trying to be too smart about it...alternatively, we could just present a file selection dialog and let the user find it themselves IF it's not in the default path. Thoughts? This might just invite a lot of questions of "I don't know where DayZ is installed," though... ( I can see it now)

> The only thing that I can think of that I did different, was to download steam via the built in Mint Software Store, rather than getting it from steam directly on their website.....should I consider nuking the installation and grabbing steam official?

No, always favor your distribution's package over a download for the purposes of linking to local libraries.

> "Searching for alternate DayZ path (make take some time)....should that be "may take some time"?

Thanks, fixed the typo.

**Comment by SteveIDusa** [comment]
### @aclist 

Yeah, that is the one as far as the dialog box, so are you telling me I should not be inclined to spend extra time making sure it will look correct for what ever distribution I use cause it will likely not matter on another???  If that is the case, how is it not standardized across distro's......I have been using Linux for over a decade easy, is it that hard to get ever distro on the same page regarding something I would consider a no brain-er?   I would think the community no matter how diverse would welcome some kind of design standardization....that does not happen?

Regarding, the version of steam and from where I get it, after posting that, I nuked that installation and tried the one from steam directly, but it just used the same paths, so not sure that it really mattered.  They both launch steam and I can see my game libraries, so they both seem to work.

I gave the search for alternate path box nearly 1/2 an hour....but I have 3 SSD's installed,  1. NVMe 1TB, 2. SATA SSD 1TB and 3. 120GB SATA SSD.  Is it set to search the boot drive first?  And on a clean install with the only things installed besides the default OS being Dropbox, which I only use the free version, so it is not very large...would/should it take more than 1/2 an hour? 

I guess I just have to let it sit and try, cause it will not create a config file until it does....

**Comment by aclist** [comment]
> Yeah, that is the one as far as the dialog box, so are you telling me I should not be inclined to spend extra time making sure it will look correct for what ever distribution I use cause it will likely not matter on another???

I think it's mostly consistent, but this dialog system uses the local system's GTK settings, so depending on what window manager/skin/theme/font size/etc. you have locally defined, your windows could look vastly different from mine. This shouldn't be a problem, but the GUI parameters expect hardcoded dimension values (which is bad), rather than just letting me specify relative numbers like 50% of screen, etc. We started by taking the hardcoded 1200x800 dimensions of the Steam Deck and working around those. Basically, we can't create anything so large that it'll become occluded on the Steam Deck, so that's the weakest link.

>  If that is the case, how is it not standardized across distro's......I have been using Linux for over a decade easy, is it that hard to get ever distro on the same page regarding something I would consider a no brain-er? I would think the community no matter how diverse would welcome some kind of design standardization....that does not happen?

Linux is all about 14,000 ways of doing the same thing. Embrace the fragmentation, baby!

> Regarding, the version of steam and from where I get it, after posting that, I nuked that installation and tried the one from steam directly, but it just used the same paths, so not sure that it really mattered. They both launch steam and I can see my game libraries, so they both seem to work.

By local libraries, I was actually referring to the equivalent of Windows .dlls, not game library.

> I gave the search for alternate path box nearly 1/2 an hour....but I have 3 SSD's installed, 1. NVMe 1TB, 2. SATA SSD 1TB and 3. 120GB SATA SSD. Is it set to search the boot drive first? And on a clean install with the only things installed besides the default OS being Dropbox, which I only use the free version, so it is not very large...would/should it take more than 1/2 an hour?

OK, yeah, I can see how this would be a problem. It's probably just going through every one of them. It starts at '/` (root path) and searches everything, including drives below that. I'm not opposed to showing a file dialog. Maybe we could attempt to find the path automatically and give up after a couple minutes? 

Some options:

1. Check for default path, failing that, ask user to select path from a file picker
2. Check for default path, failing that, ask user to enter path by hand
3. Check for default path, failing that, search the whole system but let the user interrupt this process if they get bored and select their own path
4. Check for default path, failing that, search the whole system but give up after N minutes and force the user to select the path

Any preference here?

**Comment by SteveIDusa** [comment]
### @aclist 

I had the same thoughts, I think 1 or 2 would be a good alternative or even 3, actually all 4 would be fine.  I guess I would as a person that did production manufacturing for many year and was brought up in the idea of lean manufacturing, I would suggest what ever is the least amount of work for you but provides the option to interrupt or interrupt it auto-magically after a set amount of time is how I would focus my time.  Any user that is attempting to use this is likely a little more advanced than a Windows user and should be able to do a google search to get an idea as to where to look.

I have the most recent stable version of DZGUI doing a path search now, it is into it 15 mins by my phones stop watch and counting.

**Comment by aclist** [comment]
Sorry to make you wait, but the fix was taking over an hour anyway. I basically have it working now (file picker etc.), but had to add a slew of error handling and additional control flow for if the user breaks out of the auto-search process prematurely. Not quite finished yet. Managing subprocess IDs is a bit of a pain when doing everything by spawning GUI windows versus a pure terminal application.

**Comment by aclist** [comment]
> I did save an older version, which I can implant into the DZGUI location, so as to prompt to download a newer version if you get an alternate out, until then I will just let it search and watch some vids or something.

By the way, you can just edit the first few lines of the file and change the version number to anything else if you want to trigger a version conflict. No need to insert new files.

**Comment by SteveIDusa** [comment]
### @aclist 

No problem, it finally went, I will do some more investigating tomorrow, did get a couple of hours of play time in, so it is all good!

**Comment by aclist** [comment]
I was having one of those "going down the garden path" moments with the file picker idea. I spent a couple hours elaborating it, doing the path discovery stuff and then checking if the user had canceled the process or it had failed of its own accord (giving it a 45 second cutoff), then capturing the pid of the progress window and killing it, sending the user to a file picker dialog, blah blah blah. It was getting rather finicky to deal with the process IDs of these dialogs getting spawned and going away all the time, which is really the Achilles' heel of a GUI like this where everything has to be nested and recurse over and over again. 

Anyway, I got up and walked around and asked myself, "What the hell are we actually trying to achieve, here?" The problem is dedicating too much time to scanning the directories, after all. I'll plead the insanity defense on this one.

So I attacked it from a different approach:

1. Firstly, be more permissive in checking the $HOME path for different permutations of the Steam installation. Originally we were treating $HOME/.local/share as the canonical path, but it's easier and faster to just check $HOME recursively for anything on the first pass, which will capture atypical locations like Debian's `.steam/debian-installation`.  And we don't have to add a bunch of handling for every distro.
2. Streamline the path discovery itself to skip unwanted/unpermitted directories like /var, /proc, /tmp etc where there isn't going to be an installation anyway.
3. Do away with the whole process of collecting multiple DayZ installation paths and printing a dialog (this was buried in the code just in case someone had concurrent installations) and just use a fast approach to grabbing the first and most likely one to speed up searching. If it's still somehow the wrong path, they can set it elsewhere on their own (I added some boilerplate code for a function that would let you re-set your path via a menu option if need be). This cuts out a lot of unnecessary menus, error handling, waiting on input, etc.

So first pass checks $HOME and should capture 99% of likely installations. Second pass checks `/` but skips all of the garbage and hopefully chugs through it much faster. 

I noticed a considerable uplift in scan time on my system using this new method.  I mean, basically it is instantaneous now. It takes 0-5 seconds to finish, and that's for a non-standard location on a second drive.

But I'm curious to see performance on your server farm full of hard drives. You can test the following:

Checking $HOME on the first pass:
`time find $HOME -type d -regex ".*/steamapps/common/DayZ$" -print -quit 2>/dev/null`

Trying the root path as a fallback: 
`time find / -type d \( -path "/proc" -o -path "*/timeshift" -o -path "/tmp" -o -path "/usr" -o -path "/boot" -o -path "/proc" -o -path "/root" -o -path "/run" -o -path "/sys" -o -path "/etc" -o -path "/var" -o -path "/run" -o -path "/lost+found" \) -prune -o -regex ".*/steamapps/common/DayZ$" -print -quit 2>/dev/null`

These should return a total time to process in seconds.

Next, try this without suppression of error messages:

`time find / -type d \( -path "/proc" -o -path "*/timeshift" -o -path "/tmp" -o -path "/usr" -o -path "/boot" -o -path "/proc" -o -path "/root" -o -path "/run" -o -path "/sys" -o -path "/etc" -o -path "/var" -o -path "/run" -o -path "/lost+found" \) -prune -o -regex ".*/steamapps/common/DayZ$" -print -quit`

It might return that some directories are "permission denied." I included most locations in the exclusions list, but there may be some others that could be excluded to speed up processing still further. We're basically skipping looking inside these directories altogether because it's pointless to do so or because we don't have permission. 

If you see any permission denied directories that should be added to the exclusions list, you could mention them here, unless it's something sensitive.

It took 0.537s to scan the entirety of `/` for me and find the DayZ path.

I'm pushing these changes to testing as well, but can further refine it based on your input.

**Comment by SteveIDusa** [comment]
### @aclist 

I really enjoy hearing you think, and consider.  It really does help ones brain to rewire.....when you get to see someone else doing what you do!  Thank you for sharing the though process and your conclusions.

Ran all three commands and the highest time I have using them is 110 ms, so yeah that is what I would call instantaneous!

Got up this morning early, and while waiting for the rest of the family to get moving, jumped back to DZGUI and toggled the branch.  Using 2.4.2-rc.6, when I select "Launch Server List" I am getting a blank screen.  Toggled both debug and normal modes and still no servers are listed.  When I attempt to add a server in normal mode, it provides the dialog box, and when I put in the Server ID# it then provides the "Added 13193240 to:  path which is /home/MyUserName/.config/dztui/dztuirc If errors occur, you can restore the file blah blah blah - dialog box.  I then hit the O.K button and it takes me back to the main screen, however when I then select Launch server list again it again is blank giving no optional server to choose.

I then switched over to debug mode and attempted to add a server, but to the same result, so it seems that there is bug in 2.4.2-rc.6  If I switch back to release branch, it downloads 2.4.1, creates a config file, and then displays my previously selected server as if I never toggled the branch.  I have yet to test if it will launch and allow me to play this morning, but I have other things to take care of before I can test that part.   But I thought I should let you know about the bug....at least for now.

**Comment by aclist** [comment]
Well, at least the path discovery is fixed.

OK, damn, seems we have a regression. Looks like I am working overtime. I've just confirmed the behavior and will try to get a hotfix or a rollback pushed out.

**Comment by aclist** [comment]
Should be fixed now on `2.5.0-rc.2`

**Comment by SteveIDusa** [comment]
### @aclist 

2.5.0-rc.2 does fix the no server list issue.  Going to load up the game and join the wife in some online play.  Thank you very much for you hard work!  It is Sunday however, I could have waited, all of us could have waited.  It is ok to take a breather you know???

**Comment by aclist** [comment]
OK, good. Well, there are some people taking refuge from rare bugs in the Stable branch like trying to load very large server lists or servers with a large number of mods, so I feel like they need at least one viable branch here while I hold back the release a bit more.

**Comment by SteveIDusa** [comment]
### @aclist 

Thank you again for the hard work!  It is much appreciated!  And, the other day when I did the full system wipe and reinstall of the OS, I had the reason to go back to protondb and find the Map Size command thinking I still needed it.  Noticed a mention or two, so I did a page search for DZGUI and got quite a few hits, so you are definitely filling a gap for us Linux Users!!!!  Any change to analytics?  Are you seeing it grow?  Can you see it grow?

**Comment by aclist** [comment]
Yes, I try to check that page from time to time to see if there are any issues users are facing and think about ways of preemptively addressing them.

As far as discoverability goes, this project lives inside the original DZTUI repo, so the term "DZGUI" doesn't get indexed by search engines at all, so it's still hard for people to find. I'll move it into its own repository eventually. 

After today's release and after adding connect-by-IP functionality in the release after that, I'll make a post about it on forums to give people a heads-up of where we have come since the original version. I think being able to connect to some arbitrary IP a friend gives you is still highly sought-after.

I did a feasibility test for having a server browser directly inside of the application. That is, search for and browse all of the available servers out there without having to browse them on the web or add them to a list first. It seems doable now, although we are limited by the GUI framework a bit. Still, I think it can be done, but will just take some time--and I'm not sure what if it's the "must-have" functionality right now. It's a toss-up between auto mod installation and server browser, but if we had both, we'd be sitting pretty.

I'm also developing more of a general purpose launcher for Linux games that I'll have you beta test before long, if interested.

**Comment by SteveIDusa** [comment]
Wow! Yeah, if both server browsing and mod download/loading was in a single package, you would be on par with the native launcher.  That would be very cool!  

A general purpose game launcher for Linux?  So like a Lutris, or Game On Linux type of thing or an alternate interface for something between the user and Steam?  I am curious...

**Comment by aclist** [comment]
Sort of a general interface for launching games and configuring their settings on the fly in a way not dissimilar from what is done here. I'll invite you to the project when it's ready.

As for DZGUI, connect-by-IP is now live on the Testing branch.

Requires an API key from here: https://steamcommunity.com/dev/apikey

When you choose to connect by IP, it will prompt you for API key if it's the first time. IPs can be entered as-is with no additional port. If there are multiple servers behind the same IP and running on different ports (e.g. hosting different maps), it will return them all in a list and let you select the one you want. It also prints the server metadata before connecting so that you can be sure the IP you entered is what you wanted.

**Comment by SteveIDusa** [comment]
You might be in luck, I happen to have a family member that has been hosting a local DayZ server, but they do not know how to setup port forwarding, so they and their immediate family are the only users on their local network.  Perhaps I should assist them with figuring out the port forwarding so I can test that feature......The only real problem is that they are apart of the reserves and I never know when they are home or deployed....lol

**Comment by aclist** [comment]
Actually, someone else was also asking about the possibility of connecting to LOCAL servers from within the app. I guess you still need to set up launch params even if playing locally on your own server. I'm curious to know more about how that usually happens. Do they connect to an internal IP (192...) or just "localhost"? The big question mark here is where we would pull the modlist from if joining a local server. I need more info about where these settings are stored, and this seems to be arcana only known by server operators. 

You can test connect to remote IP by simply copying the IP listed on any BM server page.

e.g. https://www.battlemetrics.com/servers/dayz/8789747 => IP given is `135.148.136.111`

This exposes multiple maps on different ports.

**Comment by SteveIDusa** [comment]
### @aclist 

Ok, had a chance to add by IP the other day, it asked me to provide a Steam API key, not having one or the time to figure out how to get one, I canceled and joined a game by server list.  Today, having forgotten the previous attempt, tried again to join by IP, but this time, it indicated that it had saved a Steam API key, not sure how and then attempted to add by IP again.  This time it gave an error that it could not add by IP with a "Failed to retrieve IP metadata, Check IP or API key and try again." error.  So at some point I have to get the Steam API Key together, but thought I would share that it for some reason thought it had one.  Server IP I attempted to use was 178.239.172.24 and the DZGUI version I am currently running is 2.6.0-rc.1

**Comment by aclist** [comment]
Well, it's not currently possible to add an empty API key in that field. If you cancel or submit an empty field, it will just take you back to the main menu and not write anything to the config file. 

It sounds like you probably hit spacebar or typed a few characters and then canceled?  I could add additional validation here to make sure the key is a certain length and at least resembles the correct API key format. 

You can check your config file in $HOME/.config/dztui/dztuirc and the `steam_api` field to see if something got inserted into there.

The process for registering is here: https://steamcommunity.com/dev/apikey

**Comment by SteveIDusa** [comment]
### @aclist 

Yep, sure did, must have misread the dialog presented and thought it was asking for the IP address, cause that is what I found in the config.  To undo this, just delete what is between the " "s or set it to a "0"?  In other words, what do I do to make it start asking about the Steam API key again?

**Comment by aclist** [comment]
1. I could add some validation to check the length of the string provided and make sure it's valid.
2. Yes, you can just delete what's between the quotation marks.
3. It's not in the documentation yet; usually I only document live features.
4. The domain parameter is a standard question given by API registration forms in the event that the application you are developing will actually callback to a specific URL (e.g., some front-facing service). If the key is just being generated for personal use, it's fine to put any generic placeholder here, such as a local IP like 127.0.0.1.

**Comment by SteveIDusa** [comment]
### @aclist 
Got a chance to get that Steam API, got it into the config, and confirmed that Add by IP did work and found 2 maps on the same ip on offer.

However, it seems I am back to an issue where I am getting missing .pbo errors.  And for some reason on this particular server, it keeps telling me that the launch conditions are satisfied and launches the game without getting any mods.  I even tried going back to the stable branch to see if that would fix it. But that did not resolve it either.  I can find it in battlemetrics as ID 12270609, I can copy and paste in the ip from BM and it shows the two, but getting actual mods to display as needed in the browser is not happening at all.   I even cleared the mods in the folders and took out all of the @(Mod Name) files. but I can not get this particular server to show the download links.

**Comment by aclist** [comment]
Do you remember how the BM site was returning malformed mods at the top of the list?

Bohemia started obfuscating mod information over the server query protocol and sending it back encoded in a special format. I escalated the issue to BM and repeatedly begged them to acknowledge the issue, but, admittedly, it was pretty complicated, because Bohemia hadn't released any specifications about how the info was encoded.

During that time, we basically swapped to using another third party API (that was decoding the info sort of correctly) as a fallback for getting the modlist.

But that other API isn't exactly working in all cases, it seems. So we are just flip-flopping between different APIs trying to get the optimal result.

Since BM seems to be working as intended, I've restored the legacy code using the original method, and it should work on the server you stated. But do please test on other servers as usual and make sure that it's reliably returning the right info.

While I would really like to develop a native method without third party APIs, I don't really have the time or resources to quit my job and create that right now...it's a non-trivial undertaking. It would be nice to have something more robust that is not subject to the vagaries of third party APIs, but it's a big task.

**Comment by SteveIDusa** [comment]
### @aclist 

Oh, yeah, it may not be related to something I am doing or not doing.....OK.  I will not pretend to even understand at this point how you are getting all of this information together, but I totally understand priorities need to be in the right place.  So I will find these changes in the stable or testing branch?

**Comment by SteveIDusa** [comment]
### @aclist 
Just tried another join by IP with 135.125.189,121....Which in this case is a JohnMcLane's official (New Map) ALTERIA v.0.3 map and I suspect also a private server.  It was announced this morning that he was opening it up for testing just for today.  Although I can see it, even using the testing branch version of 2.6.0-rc3 it seems that it is doing the same thing as last night with the other server.  Rather than opening the web browser with links for download, it just says the launch conditions are satisfied and then launches the game to a .pbo error.

I begin to suspect it has something to do with it being a private server rather than a rented server, but I may be missing other things to consider.  I have such a limited knowledge of servers, but last night I know for a fact it was a private server as the owner was discording with my wife and myself and acknowledged that it was sitting just a few feet away from them.  And I can not believe that someone that is designing some of the most popular maps for DayZ would be interested in paying a fee every month to do development, I mean perhaps, but why if you can do a private server with an older computer right in your home.

**Comment by aclist** [comment]
Thanks for the additional info. That's a regression. I had neglected to update the connect-by-ip method while updating the other part. With connect-by-ip, you aren't providing any BM ID, so we have to use the other API.

Both methods should be working now. I'm not thrilled with depending on two different things for what should be essentially the same action (connect to a server), but I guess this whole project is just a series of workarounds anyway. It pains me to think we are resting on a house of cards, though.

I can confirm that 135.125.189.121 is working when using connect by IP on `2.6.0-rc4` and not adding a BM ID.

How is the new map?

I will continue investigating the Bullets and Beans issue.

**Comment by aclist** [comment]
I'm not seeing In-Vehicle Inventory reported in the modlist reported by the server. That's one that doesn't get downloaded and then the game client reports that it is missing?

**Comment by SteveIDusa** [comment]
### @aclist 
Right, it does not seem to get added to the list, and then when I do and subscribe manually it seems that DZGUI does not know that it exists, so that causes some type of conflict.  I know it is there, I can see it, but for some reason it does not get loaded by DZGUI.  For my wife on Windows 11, she had to also do a manual subscribe, but the DayZ Launcher allows here to go into the mod list and check a box to say load it.  Of course that is not available in DZGUI, and loading up the DayZ launcher in Linux causes all of the subscribed mods to show as broken, so even if I tell it to load that mod, DZGUI still can not see it I guess.....

New map is shooting gallery at the moment.  Spawn is highly contested, got killed a lot with little loot.  After 4 or 5 hours of play we finally got it figured out that the best option is to run the coast and only hit houses for food and keep moving.....

**Comment by aclist** [comment]
Well, DZGUI has to parse the list of mods reported by the server as a single source of truth and prepare them that way. We can't just guess which mods might be correct or the user wants.

Found the problem: this one isn't showing up because the mod creator did not configure it correctly.
https://steamcommunity.com/sharedfiles/filedetails/?id=2472500243

See all of the comments posted at the bottom. The `meta.cpp` contains a 0 instead of the correct UUID. Mod ID cannot be zero. This is consistent with the erroneous results we are getting back from the API. 

The correct mod ID should be 2472500243.

It looks like the mod creator is unresponsive to this and has not fixed it.

**Comment by SteveIDusa** [comment]
### @aclist 
Thank you for doing the research to find the issue.  I guess that server is off limits unless the owner wants to change the mod list.  
But hey you have to do what will work and I for one am very happy with the results 99% of the time.  Thank you again.

**Comment by aclist** [comment]
Unfortunately, there is no standardization with servers; it's a largely homebrew effort. Mods become discontinued or unsupported by the creator and are left in limbo. I'm not sure why it's possible to upload a mod with an empty ID, but I guess there is no integrity check performed. Ultimately, it's the responsibility of the server operator to audit these. I would suggest notifying the server operator that the mod is broken.

**Comment by SteveIDusa** [comment]
Already done, and they spoke to their small community and they agreed to remove it,  Got several hours on last night.  Is it possible to alter the meta.cpp file and get it to work?  If so, I will try to talk the owner into making the change and see if that prompts me to get the correct mod.

**Comment by aclist** [comment]
I was thinking about that. If the server operator updates the ID to the correct one, it should report back correctly. Problem is, when you download the mod from Steam, it's still going to contain the wrong ID, so it's not going to be able to reconcile the local mod with the one the server expects unless we make an explicit exception for this mod ID and hardcode it to overwrite its ID. Which is bad practice, since we'd just be manually creating an exclusion for this particular mod, which wouldn't prevent the problem if another mod is improperly configured in the future.

**Comment by aclist** [comment]
One other thing you could do that would be much more effective here is for either the server operator or you to create and publish a duplicate of this mod with a correct ID, and start serving that one instead. It's easy to do. It seems that the author has abandoned the original one, along with other mods, so I wouldn't personally see that as a problem.

**Comment by SteveIDusa** [comment]
### @aclist 
I do not see anyway of doing that, do I just use the one that I subscribed too and repackage it?  Sorry I have absolutely no experience with doing anything with mods except subscribing and using them.  Where do I look to get the correct information on how to do it?  Youtube, Google, ask you?

**Comment by aclist** [comment]
I have not done it for DayZ specifically, but have uploaded mods for other games. Generally the game ships with some interface that lets you push them directly to the Steam workshop and they get registered under your account. These are trivial to remove if something was done incorrectly. There are 20,000 mods for DayZ, and they get added and deleted all the time. 

There is an application called DayZ Tools included with the DayZ license that is used to package mods. You would have to repackage the assets. This user on YouTube seems to have a whole playlist with different tutorials on doing it: https://www.youtube.com/watch?v=SPyYHHJR6oc&list=PLsFzoDww-uLyhL05aXaIA9DuqceJnrjwZ

It doesn't show up in your Steam library, but gets installed into "DayZ Tools/Bin/Launcher/DayZToolsLauncher.exe".

I haven't tried any of this myself, but it seems straightforward enough. I could maybe take a look at it when I have time.

**Comment by SteveIDusa** [comment]
Wonder if it will run under proton.....

**Comment by aclist** [comment]
You don't need the whole package, just two tools called DSUtils.exe and Publisher.exe. First one is used to sign the mod and the second is used to upload to Steam. I was able to get them working in Proton after a bunch of fiddling with installing the Microsoft Visual C++ Runtimes 2010, 2013, and 2019 into the Proton prefix. Steam needs to be running in the same Wine prefix as the tool for it to work. It takes a lot of fiddling (this used to be standard operating procedure for playing games in Wine), but is doable. 

Just doing it on Windows would obviously be faster, though. Looks like you don't need to really repackage or change anything around, just update the `meta.cpp` file and sign and publish the pbos.

PITA but it is what it is. Just offering you alternatives here.

**Comment by SteveIDusa** [comment]
LOL, I finally dropped windows because Proton could launch and let me play a few of my favorite games with out fiddling with Wine or "Play on Linux" or Lutris.  I have never had much luck with them....I guess I just am either missing prerequisite knowledge or understanding to fiddle my way through it. I will admit I once got the Epic Game Launcher installed with Lutris while GTA V was free on their system.  Played for two hours and never picked it up again....not my style I guess.

I would be totally lost with setting up the MS stuff to run in the same prefix.....I get it can be done, but that is way beyond my current knowledge.  It is kinda funny how it works too, I've noticed over the years, it is not really that something is being held secret from me, as much as I have to have the right experience at the right time with the right need or intent and suddenly I have a key to some knowledge that I did not have 3 years prior.  Because I have tried reading tutorials and following video tutorials and I still just get lost.

I am however going to get my wife to take a look at those videos you linked, since she is still using Windows, and wants to make her own map....Maybe between her and I we will get it figured out...Thanks

**Comment by aclist** [comment]
I totally agree! Sometimes you have the right knowledge at the wrong time, or the right timing and the lack of knowledge. It's best to not try to force an issue, and just let it percolate in your brain. Then at some undefined point in the future, you find you have the right skills at the right time and it's an organic process built up through steady accumulation and exposure to new ideas. That's basically a fancy explanation for what we call "experience..."

I'm going to take a crack at either the server browser or mod downloads next and will let you know when a proof of concept is ready.

**Comment by aclist** [comment]
Stealth pre-release. Full server browser is now working. No setup required, no ID pre-entry or looking up servers in your browser. Added about 200 LOC plus some helpers to calculate radius of the Earth's surface and distance between latitudinal and longitudinal points. (Haversine formula)

You can filter by map, full/empty, daytime, nighttime, non-ASCII names, low population servers, and perform keyword search.

If you untick "All Maps," you will be given a list of maps to choose from, and it will return the full list of servers hosting that map.

Non-ASCII will suppress servers with fancy characters in the name (this also affects CJK languages and Cyrillic).

Keyword search lets you search for server(s) by name. This greatly speeds up narrowing down servers--for example, if you wanted to search for the "Bullets 'n' Beans" server, you could just type "beans" or "bullets". Also really helps if looking for servers with particular rulesets like "1PP" or "fresh wipe". Search is case insensitive and can be anything.

If no servers match the filter criteria, it will return to the main menu.

The server list gets loaded in and updated in real time. There are 20,000 servers broadcasting, but only about 3,000 of them active at a given time. Using proper filters to trim down your incoming list to <100 is the smart play here, but ~500 is still manageable. No pagination is needed. You can still load 3,000 if you like, but it takes awhile to update them all in place.

Also included is geolocation to show servers by proximity. This is the best sort criterion for sorting servers geographically proximate to you. Unfortunately, this is where we are running into the limitations of an unfunded project:

We are using a free geolocation database subset to calculate latitude and longitude. This database does not contain every IP on Earth. Currently, about 40% of the servers might come back with a distance of "unknown." This is because their IPs are not found in the database. This is bad because distance is the best predictor of whether a server will be good to play on (is a server in Germany or Djibouti?), so not having that info for every single server is not ideal. Secondly, in order to work around these unlisted ones, we have to do additional, computationally expensive pre-processing of the data to find nearest possible IPs, and this is slowing down the loading of the server list. If we could reliably check every IP without these workarounds, the whole table would load faster, and with 3,000 or more entries to deal with at times, every second saved is important here.

It seems it would cost about $400 for a one-time license to obtain a complete list of geolocation coordinates, with possible annual fees after that if we wanted to constantly ensure that every IP is up to date in the list on a rolling basis (to account for new servers coming up). Otherwise, getting the full list once and not updating it for the foreseeable future would still provide pretty good coverage. 

Going forward, it looks like this server browser can replace all of the original functionality that depends on BM IDs and just let you search for/add servers on the fly. This would let us dispense with pre-registration and a lot of the finicky details. Currently, the new server browser lets you search and connect only. In order to keep a running list of servers you connected to and populate your favorites list via that, I would have to chuck out some of the legacy code and rewrite it to point to this new browser, as well as allow migration of old server lists into the new method under the hood. This should all be pretty OK to do, but it is a non-backwards-compatible change, so would bump the version to 3.0.0 at that time.

**Comment by aclist** [comment]
I forgot to mention that this now adds support for official servers.

**Comment by SteveIDusa** [comment]
### @aclist 

I got the notice of the change and tried early yesterday evening, but the browser server attempt ended in a "This requires Python" error and went back to the main navigation page.  So I reverted back to the tried and true "Known" Server list and played for a bit.  I kept having an issue on the server though that would require a disconnect and reconnect after a particular action.  If I avoided that action the issue did not happen, only when I did that action.  However this did mean that on an occasion my DayZ did a crash to desktop which meant restarting with DZGUI.  I then got the new notification for the new new version, but I had disconnected at a place that I was not feeling safe, so I did not get a chance to retry.

At first I thought to myself, "Ah, now your just showing off!!"  LOL, when I started to read the post above.  World Radius?  Geolocation?  Why?  Then of course I read more and realized what you were really doing is trying to get users the shortest distance to them or nearest server!  HOLY CRAP!  Really?

I want to make a suggestion, Even going to a full new version, Keep a "Legacy Process" link in the menu, for anyone that might find the old system useful.  Because it is, and frankly though finicky, is still pretty darn concise.  It is also a view into the process that you have undertaken and for some of us a (If murky) a little window into the process.

I cant wait to try the new browser though!  WOW!

**Comment by aclist** [comment]
OK, first update, I managed to devise a more performant and accurate way of getting geolocation matches, so brought the inaccurate hits down from 40% to about 3% of all servers. I'm building an updated DB of the missing IPs by just adding them explicitly in order to bring it up to 100% coverage (as of this date).

Secondly, managed to improve the search algorithm itself so it's pretty speedy when doing the distance check.

As to the Python thing, well, I'm glad I included that warning  as an afterthought. Python should be installed by default on almost all distributions, so I originally thought it would be fine to just drop you right in there, but I had an inkling it might be missing on certain setups. Here's what you need for your distro: https://linuxways.net/mint/how-to-install-python-on-linux-mint-20/

Other than that, no setup necessary.

As for features, yeah, I have to think about how to manage them. I don't mind keeping stuff, but the main menu is getting pretty busy. And I figured if you use the server browser, it would also passively build a "favorites" list of servers you connected to before, so you wouldn't have to add those by yourself. But I guess there might be someone out there who wants to add servers by ID and browse the web site...ya never know.

It's looking pretty great over here and provides a great way of discovering different servers and maps that are out there (there are 65 unique map types being hosted), so curious to see your feedback.

**Comment by SteveIDusa** [comment]
### @aclist 

Python 3 does indeed ship with mint by default and when we started the Python Course here at home with the kid, I was the only one that did not have to download and install it, while they on Windows did.  I have noted, that on Windows they do not have to launch a python program with anything other than the name.py on the command line.  While I on Linux Mint 20.3 and now after my reinstall Mint 21 have to launch python programs with "python3 name.py" at the command line.

I wonder if this is the crux of my systems result.  I just tried it again, and got the same python requirement.  As I have my lovely wife waiting for me to join I will have to come back to this a little later to review the update.  Still looking forward to seeing all of the improvements.

**Comment by aclist** [comment]
OK, I am going to drop the Python requirement and use a small C program instead. The requirement was only for the distance calculation, not for the table itself.

**Comment by aclist** [comment]
OK, provided I did not screw anything up, the latest update should work out of the box and gives a 25% performance uplift over the Python version. I also added some extra error handling and added a method for the client to fetch the latest geolocation tables if they are out of date. Currently I have coverage for about 19000 servers with all distances accounted for.

**Comment by SteveIDusa** [comment]
### @aclist 

Ok, just downloaded 2.7.0-rc.7.  Attempting to launch the  "(NEW) Server Browser" option resulted in the same "Requires Python" error.  Do I need to delete a config file or something?  Thought it said it wrote a new one.....?

**Comment by aclist** [comment]
Sorry, I'm dumb! I forgot to remove that check despite changing the rest of it. This is why we have testing releases...

Should be fixed now.

**Comment by aclist** [comment]
Sorry, I must be running on little sleep. After I told you I had fixed it, it looks like I didn't actually commit the code. It was still sitting on my local machine when I looked at it today. Sorry! I just bumped the version to 2.7.0-rc.8.

**Comment by SteveIDusa** [comment]
### @aclist 

OK, Thanks, I tried last night and was like "Huh?", so I was going to come and ask this afternoon and find you figured out what was going to ask about.  Will try as soon as I can!

**Comment by aclist** [comment]
Did you have any luck testing it out?

**Comment by SteveIDusa** [comment]
### @aclist 
I did try it once, but it seemed to be taking to long to respond, so I just killed the process in my System Monitor and then restarted and jumped into the game using the older process.  I had not had a chance to get back to it until this evening.

So here is where I am at., it seems to launch as expected, after choosing the browser server and selecting options, the GUI disappears and the system monitor shows 4 dzgui processes, while waiting for the GUI to come back up 2 to 3 additional un-named processes pop up at the top of the process list with ID#'s but no names.

It seems to appear like it is doing something, but never displays anything to indicate that it is progressing.
But I am not comfortable watching processes run with no progression displayed, so I kill it and move on.

**Comment by aclist** [comment]
Another question, do you usually launch DZGUI from a shortcut or from the terminal? I can add some debug logging to print messages to the terminal indicating what phase of the process we are in while it runs, as this would help identify where it's locking up. It also occurs to me that if you are using a shortcut (not that there's anything wrong with that), it might be affecting the way subprocesses are getting spawned due to the permissions of the subshell.

**Comment by aclist** [comment]
Just tested on a Steam Deck and think I may have found the issue. Can you run `sudo apt-cache policy libc6` and tell me what it reports?

**Comment by aclist** [comment]
I recompiled the distance checker against an older version of libc6 so that it supports Steam Deck and Ubuntu/Debian variants better. Seems to be working on the Steam Deck now. I did a package search for Linux Mint and I am fairly certain you are on libc6 `2.31-0ubuntu9.9` at the highest, so this change should fix support. Whether it addresses the table going missing, not sure.

**Comment by SteveIDusa** [comment]
### @aclist 

Again specs are Ryzen 3700X, 16GBs RAM, GTX 1070,

Yeah, it looks like it's trying to do something, but I am not even getting the progress bars you mention,

When you published adding it to Steam, I converted to launching from Steam.  I am getting more and more familiar with the command line, well because "Linux", but once you did that, I was "Yes! Please!"

I ran the apt-cache policy libc6 which produced this weather I used sudo or not:
libc6:
  Installed: 2.35-0ubuntu3.1
  Candidate: 2.35-0ubuntu3.1
  Version table:
 *** 2.35-0ubuntu3.1 500
        500 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     2.35-0ubuntu3 500
        500 http://archive.ubuntu.com/ubuntu jammy/main amd64 Packages

Hope all of this helps!

**Comment by aclist** [comment]
OK, that does clear things up more. I'll make a note to always test it as a Steam shortcut in the future. I assumed that doing that through Steam Deck would suffice, but there seem to be some differences.

Looks your version of libc6 is pretty new, so no problems there. I noticed that the Steam Deck was running 2.33, so I recompiled the geolocation code to ensure it works on older versions.

As for the bug, I believe (touch wood) I found the issue. When I added it as a non-Steam game using the `dzgui.desktop` file located in `$HOME/.local/share/applications/dzgui` (I assume you added this one), there was a dumb bug where a `│` character in the header was blocking creation of the table. I should have remembered that we had problems with this character before. Once I changed the text in the header, the issue cleared up instantly. I was able to troubleshoot this because I launched Steam itself from a terminal, so it was printing out debug information whenever an app was launched.

However, I found a secondary issue when exiting from the server list, where the process scanning server locations was blocking returning to the main menu, since it was waiting for all of the entries to finish getting scanned.

This doesn't happen when launching the script outside of Steam, but Steam is doing some stuff with forking processes and trying to spawn them in containers to ensure the runtime environment is sane, and it creates a situation where one thread can block the rest of the pipeline. I think "non-Steam game" was intended more for standalone applications and not something that forks and calls subshells, so it's kind of a pain here to make sure the subprocesses get cleaned up correctly after you return from each menu.

Anyway, instead of writing out the table entries in a pipe and listening for them via the menu, I backgrounded the process creating the table entries and just write them to a named pipe on a separate thread, with the table reading from those on demand. When we exit the table, the job responsible for writing the entries gets killed and stops blocking.

Thanks. I'm not exaggerating when I say that most of the functionality enjoyed by others in DZGUI is owed to you patiently testing this stuff.

**Comment by SteveIDusa** [comment]
Ok, was able to get to a server list, seems to be working as intended now, will attempt choosing a new server and testing mod downloads via the sever browser next login.  It is looking very good!

**Comment by aclist** [comment]
That's good news! Let me know.

I have the automod installation working on my local build, but need to add more error handling still.

**Comment by SteveIDusa** [comment]
### @aclist 
Did a full Browser server join today!  57 Mods and it worked quite well, with the exception of....I really wish it were automated, man that took for ever!  But it worked very well!!  Thank you again for the hard work!

**Comment by aclist** [comment]
Very nice! Yipee! 57 mods, ouch. Automated is...coming. I am testing different distributions in a VM in order to have better coverage.

**Comment by aclist** [comment]
I'd say 2-4 days for automation. I am going to release the server browser to the stable branch first. I'm thinking of just making auto-install a default option instead of a toggle, and having it available regardless of the user's settings. There will be an initial check of whether the automation routine will work. If it doesn't, revert to manual mode. I think this is OK (probably), as it allows for fast-tracking people onto automation without burying it in the settings.

**Comment by aclist** [comment]
I'm going to close this thread, as it's getting rather long in the tooth and will be hard for other people to refer to, having over 93 comments in it. I've opened this "forum" as a place to vote on issues, discuss beta testing, propose features, etc., and anything else that doesn't fit into the category of bug: https://github.com/aclist/dztui/discussions

You can use this thread specifically to talk about the testing branch: https://github.com/aclist/dztui/discussions/35

It would be best to continue this conversation over there with regard to the auto mod installation.

See you on the other side!

**Comment by SteveIDusa** [comment]
Understood, thank you again!
